docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+The disciminator is therefore placed in bit 20, leaving bits 0-19 for the payload.
+
+The `U20` uses all of the available payload bits and so is stored in bits 0-19.
+The `U16` follows the rule that it should be aligned to the end of the opposite end to the discriminator and is therefore stored in bits 0-15.
+
+A match expression first loads the entire 32-bit value and compares bits 21-31 against an all-ones bit pattern.
+If the comparison result is not-equal, the entire value is a `Float32`.
+If the comparison result is equal, then a value of bit 20 is tested, with zero indicating a `U16` and one indicating a `U20`, either of which can be extracted by a bit shift (note: On Arm systems, either value can be extracted with a single `UBFX` instruction, on x86 systems that support the BMI1 instructions a BEXTR can do the same).
+
+### Example: `Float64 | T* | U*`
+
+The payload size of `Float64` is 64 bits and it requires one discriminator state.
+The payload size of `T*` and `U*` is each either 32, 48, or 128 bits, depending on the target architecture and they require one discriminator state between them.
+The lowering for this union therefore depends on the target.
+
+On 32-bit systems, the largest type is `Float64` and the remaining types all fit easily in the significand bits.
+There are no discriminator bits required but the 'as-if' rules triggers as if there were a single bit for determining which end of the space should be used and the remaining values are stored at the end of the space that is strongly aligned: bits 0-31.
+
+On 64-bit systems, the same rules apply, packing the used bits of the pointer into bits 0-50.
+
+On CHERI systems, the pointer types are larger than the `Float64` and so the payload is a 128-bit value, distinguished by the tag bit.
+
+### Example: `T* | None`
+
+In this example, `None` is a singleton type with no fields, used as a placeholder indicating a not-present value.
+The `T*` requires one pointer's worth of space for the payload (as above, this varies depending on the target) and one discriminator state.
+The `None` value requires one discriminator state and no payload.
+
+The largest value is a pointer and so on CHERI systems the tag bit is used as the discriminator, on other architectures the low bit of the pointer is used (because the pointer is the largest type and the discriminator fits in a three-bit value).
+The discriminator values are assigned starting from the singletons, so zero is used for `None` and one for `T*`.
+
+A pattern match expression on this type is compiled into a compare-and-branch-on-{non}-zero, which is a single instruction on most architectures.
+Accessing the `T*` on a CHERI system requires no further operations.
+On integer-pointer systems, it requires subtracting one.
+On any mainstream architecture other than RISC-V, this subtraction can be folded into the immediate offset on a load instruction and so requires no additional instructions.
```

Comment:
Why isn't that true on RISC-V?  AFAICT loads and stores have a sign-extended 12 bit immediate offset; is that not sufficient?

Reply:
They might have fixed this after the v8 team yelled at them, but last time I looked the immediate field in the RISC-V load/store instructions was a multiple of the loaded value size, so you can't do a 3-byte offset on a 16-bit or larger load.  This means you can't fold the clearing of the low bits into the immediate, because it will be base + field offset - {1-7}.

This is what happens when you optimise your ISA based on the output of a prototype compiler on a small corpus of C code.

Reply:
It looks like it's been corrected.  Sail since reading the official ISA doc is an exercise in frustration.

https://github.com/rems-project/sail-riscv/blob/ae6cb1de092e9ea727e2318d76e9b88999bbee59/model/riscv_insts_base.sail#L317-L337

and

https://github.com/rems-project/sail-riscv/blob/f9f066e89a5c20f2769fbfb2b6a585af9b3f0561/model/riscv_addr_checks.sail#L49-L52

Reply:
Ah, great!  Odd that there's no mention of this change in the history bit of the RISC-V spec, given that it's a breaking change from the original.